components/board/genericlinux/hw_pwm.go

```diff
@@ -0,0 +1,174 @@
+		return err
+	}
+	pwm.isExported = true
+	return nil
+}
+
+// Unexport turns off any PWM signal the pin was providing, and tells the OS that this pin is no
+// longer in use (so it can be reused as an input pin, etc.).
+func (pwm *pwmDevice) unexport() error {
+	if !pwm.isExported {
+		return nil // Already unexported
+	}
+	// If we unexport the pin while it is enabled, it might continue outputting a PWM signal,
+	// causing trouble if you start using the pin for something else.
+	if err := pwm.disable(); err != nil {
+		return err
+	}
+	if err := pwm.writeChip("unexport", uint64(pwm.line)); err != nil {
+		return err
+	}
+	pwm.isExported = false
+	return nil
+}
+
+// Enable tells an exported pin to output the PWM signal it has been configured with.
+func (pwm *pwmDevice) enable() error {
+	if pwm.isEnabled {
+		return nil // Already enabled
+	}
+	if err := pwm.writeLine("enable", 1); err != nil {
+		return err
+	}
+	pwm.isEnabled = true
+	return nil
+}
+
+// Disable tells an exported pin to stop outputting its PWM signal, but it is still available for
+// reconfiguring and re-enabling.
+func (pwm *pwmDevice) disable() error {
+	if !pwm.isEnabled {
+		return nil // Already disabled
+	}
+	if err := pwm.writeLine("enable", 0); err != nil {
+		return err
+	}
+	pwm.isEnabled = false
+	return nil
+}
+
+// SetPwm configures an exported pin and enables its output signal.
+// Warning: if this function returns a non-nil error, it could leave the pin in an indeterminate
+// state. Maybe it's exported, maybe not. Maybe it's enabled, maybe not. The new frequency and duty
+// cycle each might or might not be set.
+func (pwm *pwmDevice) SetPwm(freqHz uint, dutyCycle float64) error {
+	pwm.mu.Lock()
+	defer pwm.mu.Unlock()
+
+	// What we really want in this function is a monad: for every interaction with sysfs, check if
+	// it returned an error, and if so return early.
+	if err := pwm.export(); err != nil {
+		return err
+	}
+	if err := pwm.disable(); err != nil {
+		return err
+	}
+
+	// Sysfs has a pseudofile named duty_cycle which contains the number of nanoseconds that the
+	// pin should be high within a period. It's not how the rest of the world defines a duty cycle,
+	// so we will refer to it here as the active duration.
+	periodNs := 1000 * 1000 * 1000 / uint64(freqHz)
+	activeDurationNs := uint64(float64(periodNs) * dutyCycle)
+
+	// We are never allowed to set the active duration higher than the period. Change the order we
+	// set the values to ensure this.
+	if periodNs < pwm.activeDurationNs {
+		// The new period is smaller than the old active duration. Change the active duration
+		// first.
+		if err := pwm.writeLine("duty_cycle", activeDurationNs); err != nil {
+			return err
+		}
+		pwm.activeDurationNs = activeDurationNs
+
+		if err := pwm.writeLine("period", periodNs); err != nil {
+			return err
+		}
+		pwm.periodNs = periodNs
+	} else {
+		// The new period is at least as large as the old active duration. It's safe to change the
```

Comment:
If you ever try to make the active duration longer than the period, you'll get an error and your change won't occur. This can be from making the period too small, or from making the active duration too long. So, there are 3 situations to consider:
1. The new active duration is longer than the old period. Changing the active duration first would result in an error, so we need to change the period first.
2. The new period is shorter than the old active duration. Changing the period first would result in an error, so we need to change the active duration first. Note that this can never happen at the same time as the previous case: the period is always at least as long as the active duration, both for old values and new ones.
3. Neither of the previous cases is true: the new period is at least as long as the old active duration, and the new active duration is at least as short as the old period. You can change either one first; you'll never have errors.

So the code is really
```
if we're in case 2:
    change the active duration first
else: // we're in either case 1 or 3
    change the period first
```

If that didn't make sense, let's talk in person. I suspect the easiest way to understand this is by drawing a picture.